File: tests/golden/08-arrays-iterators.ts
/**
 * 測試 08: 陣列操作與迭代器
 */

// 基本陣列操作
const numbers = [1, 2, 3, 4, 5];
const doubled = numbers.map(n => n * 2);
const filtered = numbers.filter(n => n > 2);
const sum = numbers.reduce((acc, n) => acc + n, 0);

// 陣列方法鏈
function processNumbers(nums: number[]): number {
  return nums
    .filter(n => n > 0)
    .map(n => n * 2)
    .reduce((sum, n) => sum + n, 0);
}

// Array 泛型方法
function first<T>(arr: T[]): T | undefined {
  return arr[0];
}

function last<T>(arr: T[]): T | undefined {
  return arr[arr.length - 1];
}

function take<T>(arr: T[], n: number): T[] {
  return arr.slice(0, n);
}

function chunk<T>(arr: T[], size: number): T[][] {
  const chunks: T[][] = [];
  for (let i = 0; i < arr.length; i += size) {
    chunks.push(arr.slice(i, i + size));
  }
  return chunks;
}

// for...of 迭代
function sumArray(numbers: number[]): number {
  let total = 0;
  for (const num of numbers) {
    total += num;
  }
  return total;
}

// for...in 迭代（物件）
function getKeys(obj: Record<string, any>): string[] {
  const keys: string[] = [];
  for (const key in obj) {
    if (obj.hasOwnProperty(key)) {
      keys.push(key);
    }
  }
  return keys;
}

// 解構賦值
function processCoordinates(coords: [number, number][]): void {
  for (const [x, y] of coords) {
    console.log(`Point: (${x}, ${y})`);
  }
}

// Rest 參數
function concatenate(...arrays: number[][]): number[] {
  return arrays.reduce((result, arr) => [...result, ...arr], []);
}

// Spread 運算子
const arr1 = [1, 2, 3];
const arr2 = [4, 5, 6];
const combined = [...arr1, ...arr2];
const cloned = [...arr1];

// Array.from
function range(start: number, end: number): number[] {
  return Array.from({ length: end - start }, (_, i) => start + i);
}

// 自訂迭代器
class Range implements Iterable<number> {
  constructor(
    private start: number,
    private end: number,
    private step: number = 1
  ) {}

  *[Symbol.iterator](): Iterator<number> {
    for (let i = this.start; i < this.end; i += this.step) {
      yield i;
    }
  }

  toArray(): number[] {
    return [...this];
  }
}

// Map/Set 操作
function uniqueValues<T>(arr: T[]): T[] {
  return [...new Set(arr)];
}

function countOccurrences<T>(arr: T[]): Map<T, number> {
  const counts = new Map<T, number>();
  for (const item of arr) {
    counts.set(item, (counts.get(item) || 0) + 1);
  }
  return counts;
}

// 高階陣列函式
function partition<T>(arr: T[], predicate: (item: T) => boolean): [T[], T[]] {
  const pass: T[] = [];
  const fail: T[] = [];

  for (const item of arr) {
    if (predicate(item)) {
      pass.push(item);
    } else {
      fail.push(item);
    }
  }

  return [pass, fail];
}

function groupBy<T, K extends string | number>(
  arr: T[],
  keyFn: (item: T) => K
): Record<K, T[]> {
  const groups = {} as Record<K, T[]>;

  for (const item of arr) {
    const key = keyFn(item);
    if (!groups[key]) {
      groups[key] = [];
    }
    groups[key].push(item);
  }

  return groups;
}

// 遞迴陣列處理
type NestedArray<T> = T | NestedArray<T>[];

function flatten<T>(arr: NestedArray<T>[]): T[] {
  const result: T[] = [];

  for (const item of arr) {
    if (Array.isArray(item)) {
      result.push(...flatten(item as NestedArray<T>[]));
    } else {
      result.push(item as T);
    }
  }

  return result;
}

export {
  processNumbers,
  first,
  last,
  take,
  chunk,
  sumArray,
  getKeys,
  processCoordinates,
  concatenate,
  range,
  Range,
  uniqueValues,
  countOccurrences,
  partition,
  groupBy,
  flatten
};

File: tests/golden/expected/08-arrays-iterators.go
package main

var (
	numbers  = []int{1, 2, 3, 4, 5}
	doubled  []int
	filtered []int
	sum      int
)

func init() {
	doubled = make([]int, len(numbers))
	for i, n := range numbers {
		doubled[i] = n * 2
	}

	filtered = make([]int, 0)
	for _, n := range numbers {
		if n > 2 {
			filtered = append(filtered, n)
		}
	}

	sum = 0
	for _, n := range numbers {
		sum += n
	}
}

func ProcessNumbers(nums []int) int {
	filtered := make([]int, 0)
	for _, n := range nums {
		if n > 0 {
			filtered = append(filtered, n)
		}
	}

	mapped := make([]int, len(filtered))
	for i, n := range filtered {
		mapped[i] = n * 2
	}

	sum := 0
	for _, n := range mapped {
		sum += n
	}

	return sum
}

func First[T any](arr []T) *T {
	if len(arr) == 0 {
		return nil
	}
	return &arr[0]
}

func Last[T any](arr []T) *T {
	if len(arr) == 0 {
		return nil
	}
	return &arr[len(arr)-1]
}

func Take[T any](arr []T, n int) []T {
	if n > len(arr) {
		n = len(arr)
	}
	return arr[0:n]
}

func Chunk[T any](arr []T, size int) [][]T {
	chunks := make([][]T, 0)
	for i := 0; i < len(arr); i += size {
		end := i + size
		if end > len(arr) {
			end = len(arr)
		}
		chunks = append(chunks, arr[i:end])
	}
	return chunks
}

func SumArray(numbers []int) int {
	total := 0
	for _, num := range numbers {
		total += num
	}
	return total
}

func GetKeys(obj map[string]interface{}) []string {
	keys := make([]string, 0, len(obj))
	for key := range obj {
		keys = append(keys, key)
	}
	return keys
}

func ProcessCoordinates(coords [][2]int) {
	for _, coord := range coords {
		x, y := coord[0], coord[1]
		println("Point: (", x, ", ", y, ")")
	}
}

func Concatenate(arrays ...[]int) []int {
	result := make([]int, 0)
	for _, arr := range arrays {
		result = append(result, arr...)
	}
	return result
}

var (
	arr1     = []int{1, 2, 3}
	arr2     = []int{4, 5, 6}
	combined []int
	cloned   []int
)

func init() {
	combined = append(append([]int{}, arr1...), arr2...)
	cloned = append([]int{}, arr1...)
}

func Range(start int, end int) []int {
	length := end - start
	result := make([]int, length)
	for i := 0; i < length; i++ {
		result[i] = start + i
	}
	return result
}

type RangeIterator struct {
	start int
	end   int
	step  int
}

func NewRange(start int, end int, step int) *RangeIterator {
	if step == 0 {
		step = 1
	}
	return &RangeIterator{
		start: start,
		end:   end,
		step:  step,
	}
}

func (r *RangeIterator) Iterator() <-chan int {
	ch := make(chan int)
	go func() {
		defer close(ch)
		for i := r.start; i < r.end; i += r.step {
			ch <- i
		}
	}()
	return ch
}

func (r *RangeIterator) ToArray() []int {
	result := make([]int, 0)
	for val := range r.Iterator() {
		result = append(result, val)
	}
	return result
}

func UniqueValues[T comparable](arr []T) []T {
	seen := make(map[T]bool)
	result := make([]T, 0)

	for _, item := range arr {
		if !seen[item] {
			seen[item] = true
			result = append(result, item)
		}
	}

	return result
}

func CountOccurrences[T comparable](arr []T) map[T]int {
	counts := make(map[T]int)
	for _, item := range arr {
		counts[item]++
	}
	return counts
}

func Partition[T any](arr []T, predicate func(T) bool) ([]T, []T) {
	pass := make([]T, 0)
	fail := make([]T, 0)

	for _, item := range arr {
		if predicate(item) {
			pass = append(pass, item)
		} else {
			fail = append(fail, item)
		}
	}

	return pass, fail
}

func GroupBy[T any, K comparable](arr []T, keyFn func(T) K) map[K][]T {
	groups := make(map[K][]T)

	for _, item := range arr {
		key := keyFn(item)
		groups[key] = append(groups[key], item)
	}

	return groups
}

type NestedArray interface{}

func Flatten(arr []NestedArray) []interface{} {
	result := make([]interface{}, 0)

	for _, item := range arr {
		if nested, ok := item.([]NestedArray); ok {
			result = append(result, Flatten(nested)...)
		} else {
			result = append(result, item)
		}
	}

	return result
}